refactor: MovieDetailsScreen

Mostafa-alsaygh opened the pull request:
# What does this PR do?

This PR refactors the navigation logic within the `MovieDetailsScreen`.

- Navigation callbacks (`onNavigateBack`, `onNavigateGenre`, etc.) are removed from `MovieDetailsScreen`'s parameters.
- The `NavController` is now directly passed to `MovieDetailsScreen`.
- The `HandleMovieDetailsEffects` composable function is removed, and its logic is integrated directly into the `MovieDetailsScreen`'s effect handling.
- `MovieDetailsEffect` names are updated for clarity (e.g., `BackNavigation` to `NavigateBack`).
- `MovieDetailsContract` function names are updated for consistency (e.g., `onRetry` to `onRetryClick`).
- `MovieDetailsViewModel` is updated to emit the renamed effects and use the renamed contract functions.
- The `movieHaveTrailer` property in `MovieDetailsUiState` is renamed to `hasTrailer`.
- The `NovixApp` navigation setup for `MovieDetailsScreen` is updated to pass the `NavController` directly.
- Minor refactoring in `MovieDetailsViewModel` for better readability and separation of concerns (e.g., extracting `handleMovieDetailsData`, `onRateClickSuccess`, `onRatingSuccess`, `fetchMovieData`, `onMovieDataLoaded`, `updateMovieDetailsState`, `addMovieToRecentHistory`, `setLoadingState`).

Mostafa-alsaygh committed:
refactor: MovieDetailsScreen

london-bot[bot] committed:
Branch was auto-updated.

Mostafa-alsaygh committed:
refactor: rename MovieDetailsEffect cases for clarity

london-bot[bot] committed:
Branch was auto-updated.

london-bot[bot] committed:
Branch was auto-updated.

Abdulrahman-Ragab-01 commented on app/src/main/java/com/london/app/navigation/NovixApp.kt:
```diff
@@ -463,24 +463,7 @@ fun NavGraphBuilder.mainNavGraph(
-            onNavigateToLogin = {
-                navController.navigate(Screen.Login)
-            }
+            navController = navController
```
make the navigation as we discuss for consistency

Mostafa-alsaygh replied:
we need to discuss it with ahmed mones

Abdulrahman-Ragab-01 commented on presentation/src/main/java/com/london/presentation/feature/details/movie/MovieDetailsScreen.kt:
```diff
@@ -423,33 +444,6 @@ private fun RatingAndMetaRow(
-
-
 @Composable
 private fun IconWithText(
```
don't forget to use the shared TextWithIcon

Abdulrahman-Ragab-01 reviewed (commented):
Add unit test for MovieDetailsViewModel

london-bot[bot] committed:
Branch was auto-updated.

london-bot[bot] committed:
Branch was auto-updated.

london-bot[bot] committed:
Branch was auto-updated.

london-bot[bot] committed:
Branch was auto-updated.

london-bot[bot] committed:
Branch was auto-updated.

BassantM commented on presentation/src/main/java/com/london/presentation/feature/details/movie/MovieDetailsScreen.kt:
```diff
@@ -88,32 +92,49 @@ import com.london.presentation.utils.toLocalizedNumbers
-    onNavigateToMovie: (Int) -> Unit,
-    onNavigateToActor: (Int) -> Unit,
-    onNavigateToReviews: (Int, Int) -> Unit,
+    navController: NavController,
```
why nav controller in ui ??!

Mostafa-alsaygh replied:
we don't need to put them in the novix app at all

BassantM commented on presentation/src/main/java/com/london/presentation/feature/details/movie/MovieDetailsUiState.kt:
```diff
@@ -23,12 +23,12 @@ data class MovieDetailsUiState(
     val isGuestUser: Boolean = false,
     val actors: List<Actor> = listOf(),
     val movieGenres: List<Int> = listOf(),
-    val movieImages: List<String> = listOf(),
     val isSuccessfullyRated: Boolean? = null,
+    val movieImages: List<String> = listOf(),
     val similarMovies: List<Movie> = listOf(),
     val isRateBottomSheetVisible: Boolean = false,
     val isGuestUserBottomSheetVisible: Boolean = false,
 ){
-    val movieHaveTrailer: Boolean
+    val hasTrailer: Boolean
         get() = movieVideo.isNotEmpty()
```
kindly, check actor details and folllow it , we dont need all those params

Mostafa-alsaygh committed:
refactor: update MovieDetailsScreen to use callbacks for navigation

london-bot[bot] committed:
Branch was auto-updated.

BassantM reviewed (commented):
Kindly, check a new composables are merged to refactor the movie details screen, u have a sample in continue watching screen
https://github.com/LondonSquad/Novix/pull/542

Mostafa-alsaygh committed:
refactor: improve MovieDetailsScreen structure and use TextWithIcon

Mostafa-alsaygh committed:
refactor: remove unused fields from MovieDetailsUiState

london-bot[bot] committed:
Branch was auto-updated.

BassantM committed:
Merge remote-tracking branch 'origin/develop' into refactor/movie-details

# Conflicts:
#	presentation/src/main/java/com/london/presentation/feature/details/movie/MovieDetailsContract.kt
#	presentation/src/main/java/com/london/presentation/feature/details/movie/MovieDetailsEffect.kt
#	presentation/src/main/java/com/london/presentation/feature/details/movie/MovieDetailsScreen.kt
#	presentation/src/main/java/com/london/presentation/feature/details/movie/MovieDetailsViewModel.kt

london-bot[bot] committed:
Branch was auto-updated.

yusufnasserdev commented on presentation/src/main/java/com/london/presentation/feature/details/movie/MovieDetailsViewModel.kt:
```diff
@@ -55,38 +57,27 @@ class MovieDetailsViewModel @Inject constructor(
 
     override fun onGenreClick(genreId: Int) {
-        emitEffect(MovieDetailsEffect.GenreNavigation(genreId))
+        emitEffect(MovieDetailsEffect.GenreMoviesNavigation(genreId))
```
This doesn't exist in the Effects, I think the naming might have been reverted by an automatic merge, kindly check this.

An appropriate rename would be `MoviesGenreNavigation`

yusufnasserdev commented on presentation/src/main/java/com/london/presentation/feature/details/movie/MovieDetailsViewModel.kt:
```diff
@@ -202,8 +160,63 @@ class MovieDetailsViewModel @Inject constructor(
+        }
+    }
+
+    private suspend fun addMovieToRecentHistory(details: MovieDetails) {
```
Why do you need this parameter while you have everything available in the `uiState`?

Mostafa-alsaygh replied:
it is not for the ui state

yusufnasserdev commented on presentation/src/main/java/com/london/presentation/feature/details/movie/MovieDetailsViewModel.kt:
```diff
@@ -202,8 +160,63 @@ class MovieDetailsViewModel @Inject constructor(
+        )
+    }
+
+    private fun setLoadingState(isLoading: Boolean) {
+        updateState { copy(isLoading = isLoading) }
+    }
```
Why not composite function?

Abdulrahman-Ragab-01 replied:
Bareq just recommend to keep it as it is now specially in view models because we can need to add more functionality in the future

yusufnasserdev commented on presentation/src/main/java/com/london/presentation/feature/details/movie/MovieDetailsViewModel.kt:
```diff
@@ -202,8 +160,63 @@ class MovieDetailsViewModel @Inject constructor(
         )
     }
+
+    private suspend fun fetchMovieData(movieId: Int): Triple<MovieDetails, List<String>, List<Actor>> {
+        val movie = movieDetails.getMovieDetails(movieId)
+        val movieImages = movieDetails.getMovieImagesUseCase(movieId)
+        val movieCast = movieDetails.getMovieCast(movieId)
+        return Triple(movie, movieImages, movieCast)
+    }
+
+    private suspend fun onMovieDataLoaded(triple: Triple<MovieDetails, List<String>, List<Actor>>) {
+        updateMovieDetailsState(triple)
+        addMovieToRecentHistory(triple.first)
+    }
+
+    private fun updateMovieDetailsState(triple: Triple<MovieDetails, List<String>, List<Actor>>) {
+        val (details, images, cast) = triple
+        updateState {
+            copy(
+                movieId = details.id,
+                movieName = details.title,
+                movieGenres = details.genresId,
+                movieRating = details.voteAverage,
+                movieDuration = details.runtime.toString(),
+                releaseDate = details.releaseDate,
+                movieOverview = details.overview,
+                movieImages = images.ifEmpty { listOf(details.posterUrl) },
+                actors = cast
+            )
+        }
+    }
```
Kindly reconsider the use of triple, it doesn't provide a descriptive naming for either what a function return or accepts.

Mostafa-alsaygh replied:
done

yusufnasserdev commented on presentation/src/main/java/com/london/presentation/feature/details/movie/MovieDetailsViewModel.kt:
```diff
@@ -96,79 +87,46 @@ class MovieDetailsViewModel @Inject constructor(
-                    )
+    private fun onRateClickSuccess(isLoggedIn: Boolean) {
+        if (isLoggedIn)
+            updateState { copy(isRateBottomSheetVisible = isRateBottomSheetVisible.not()) }
```
Why doesn't the `rateBottomSheetState` get its own function?

Mostafa-alsaygh replied:
it is only a fun when it is Success it will closed

yusufnasserdev commented on presentation/src/main/java/com/london/presentation/feature/details/movie/MovieDetailsViewModel.kt:
```diff
@@ -96,79 +87,46 @@ class MovieDetailsViewModel @Inject constructor(
+        else
+            updateState {
+                copy(
+                    isGuestUserBottomSheetVisible = isGuestUserBottomSheetVisible.not(),
```
Why doesn't the `GuestUserBottomSheet` get its own function?

Mostafa-alsaygh replied:
we dont need it, it is only a fun when it is Success it will closed

yusufnasserdev reviewed (commented):
Great effort, kindly check earlier comments from @BassantM & @Abdulrahman1890 as well.

Mostafa-alsaygh committed:
Merge branch 'refs/heads/develop' into refactor/movie-details

# Conflicts:
#	presentation/src/main/java/com/london/presentation/feature/details/movie/MovieDetailsScreen.kt
#	presentation/src/main/java/com/london/presentation/feature/details/movie/MovieDetailsUiState.kt
#	presentation/src/main/java/com/london/presentation/feature/details/movie/MovieDetailsViewModel.kt

Mostafa-alsaygh committed:
Refactor movie details screen and view model

This commit refactors the MovieDetailsScreen and MovieDetailsViewModel.

Key changes:

-   Simplified the effect handling in `MovieDetailsScreen` by directly calling navigation functions.
-   Removed redundant UI components like separator dots between movie info.
-   Updated `MovieDetailsViewModel` to directly use `MovieDetails`, `List<String>`, and `List<Actor>` instead of a `Triple`.
-   Modified navigation function calls in `mainNavGraph.kt` to match the updated `MovieDetailsScreen` signature.
-   Ensured `MovieGenreUi` is used for genre navigation.
-   Updated `MediaType` usage for reviews navigation.

Mostafa-alsaygh committed:
Refactor: Rename `getMovieImagesUseCase` to `getMovieImages`

The method `getMovieImagesUseCase` in `GetMovieUseCase` has been renamed to `getMovieImages` for better clarity and consistency. This change has been propagated to its usages in `MovieDetailsViewModel` and corresponding test files.

london-bot[bot] committed:
Branch was auto-updated.

london-bot[bot] committed:
Branch was auto-updated.
